# 🐶 Proyecto Big Data: EDA de Productos para Mascotas

## Etapa 2: Procesamiento y análisis con Apache Spark

En este notebook se realizará:

- conexión entre Spark y MongoDB Atlas,
- lectura de la colección RAW,
- análisis exploratorio de datos (EDA),
- tratamiento de nulos,
- detección de duplicados,
- creación de variables derivadas,
- preparación de datos para Machine Learning.

La colección utilizada será:

```text
ProyectoSemana9.Mercado_Mascotas_Raw

### Paso 1: Importar librerías

Objetivo:

- Cargar Spark.
- Cargar funciones necesarias para limpieza y transformación.

In [2]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import (col,when,regexp_extract,regexp_replace,lower,trim,isnan,count)

### Paso 2: Crear conexión entre Spark y MongoDB Atlas

Objetivo:

- Configurar la sesión Spark.
- Conectarse a Atlas.

In [3]:
# Crea session Spark
uri = "mongodb+srv://Finanzasymercado:ternurines123@cluster0.uia9vsi.mongodb.net/?retryWrites=true&w=majority"

spark = SparkSession.builder \
    .appName("EDA_Mascotas_Yahima") \
    .config("spark.mongodb.read.connection.uri", uri) \
    .config("spark.mongodb.write.connection.uri", uri) \
    .config("spark.jars.packages","org.mongodb.spark:mongo-spark-connector_2.12:10.1.1") \
    .getOrCreate()

### Paso 3: Leer colección RAW
Objetivo:
- Cargar los datos extraídos en el cuaderno anterior.

In [4]:
#Leer desde Mongo
df_raw = spark.read.format("mongodb") \
    .option("database", "ProyectoSemana9") \
    .option("collection", "Mercado_Mascotas_Raw") \
    .load()

In [5]:
# Verifica los datos
print("Cantidad total de registros:")

print(df_raw.count())

Cantidad total de registros:
620


### Paso 4: Exploración inicial del dataset (EDA inicial)

Aquí se deberían responder:

- ¿Cuántos registros existen?
    -df_raw.count()
- ¿Qué columnas posee?
    -df_raw.printSchema()
- ¿Cómo se ven los datos?
    -df_raw.show()
- ¿Qué tipo de dato tiene cada campo?
    -df_raw.dtypes

In [6]:
# Ver esquema
df_raw.printSchema()

root
 |-- _id: string (nullable = true)
 |-- categoria: string (nullable = true)
 |-- estado_extraccion: string (nullable = true)
 |-- fecha_captura: string (nullable = true)
 |-- formato_raw: string (nullable = true)
 |-- fuente_tipo: string (nullable = true)
 |-- marca: string (nullable = true)
 |-- nombre_producto: string (nullable = true)
 |-- opiniones: integer (nullable = true)
 |-- pagina: integer (nullable = true)
 |-- precio_raw: string (nullable = true)
 |-- rating: double (nullable = true)
 |-- responsable: string (nullable = true)
 |-- responsable_scraper: string (nullable = true)
 |-- sku_id: string (nullable = true)
 |-- texto_crudo: string (nullable = true)
 |-- tienda: string (nullable = true)
 |-- tipo_dato: string (nullable = true)
 |-- url_fuente: string (nullable = true)



In [7]:
# Ver Datos
display(df_raw.limit(10).toPandas())

,_id,categoria,estado_extraccion,fecha_captura,formato_raw,fuente_tipo,marca,nombre_producto,opiniones,pagina,precio_raw,rating,responsable,responsable_scraper,sku_id,texto_crudo,tienda,tipo_dato,url_fuente
0,6a1a05c5e7fe73afeef4c0b8,Pienso para perros,None,2026-05-29 21:33:09,5.76€ el kg,None,acana,acana adult pienso para perros,213,NaN,19.99€ a 195.98€,4.8,Yahima,None,TD_00001,-14% > 65€ patrocinado patrocinado acana adult...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
1,6a1a05c5e7fe73afeef4c0b9,Pienso para perros,None,2026-05-29 21:33:09,5.15€ el kg,None,criadores,criadores dietetic hipoalergénico adulto piens...,293,NaN,19.99€ a 123.48€,4.6,Yahima,None,TD_00002,-20% dto. criadores dietetic hipoalergénico ad...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
2,6a1a05c5e7fe73afeef4c0ba,Pienso para perros,None,2026-05-29 21:33:09,5.14€ el kg,None,nature's,nature's variety no grain adult medium maxi sa...,63,NaN,28.59€ a 123.46€,4.7,Yahima,None,TD_00003,¡kilos gratis! patrocinado patrocinado nature'...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
3,6a1a05c5e7fe73afeef4c0bb,Pienso para perros,None,2026-05-29 21:33:09,4.43€ el kg,None,criadores,criadores dietetic adulto obesity pienso para ...,241,NaN,18.99€ a 106.38€,4.6,Yahima,None,TD_00004,-20% dto. criadores dietetic adulto obesity pi...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
4,6a1a05c5e7fe73afeef4c0bc,Pienso para perros,None,2026-05-29 21:33:09,6.31€ el kg,None,acana,acana light & fit pienso para perros,277,NaN,21.89€ a 143.84€,4.8,Yahima,None,TD_00005,-14% > 65€ patrocinado patrocinado acana light...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
5,6a1a05c5e7fe73afeef4c0bd,Pienso para perros,None,2026-05-29 21:33:09,7.51€ el kg,None,hill's,hill's prescription diet j/d metabolic + mobil...,377,NaN,55.69€ a 180.30€,4.8,Yahima,None,TD_00006,-14% > 65€ hill's prescription diet j/d metabo...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
6,6a1a05c5e7fe73afeef4c0be,Pienso para perros,None,2026-05-29 21:33:09,9.15€ el kg,None,nature's,nature's variety no grain puppy mini salmón no...,2,NaN,5.49€ a 29.09€,5.0,Yahima,None,TD_00007,-30% en 2ª ud. patrocinado patrocinado nature'...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
7,6a1a05c5e7fe73afeef4c0bf,Pienso para perros,None,2026-05-29 21:33:09,1.27€ el kg,None,salvaje,salvaje base adultos pollo pienso para perros,391,NaN,8.99€ a 37.98€,4.5,Yahima,None,TD_00008,¡pack ahorro! salvaje base adultos pollo piens...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
8,6a1a05c5e7fe73afeef4c0c0,Pienso para perros,None,2026-05-29 21:33:09,6.86€ el kg,None,advance,advance veterinary diets hypoallergenic pienso...,43,NaN,30.19€ a 137.18€,4.3,Yahima,None,TD_00009,¡pack ahorro! patrocinado patrocinado advance ...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...
9,6a1a05c5e7fe73afeef4c0c1,Pienso para perros,None,2026-05-29 21:33:09,4.43€ el kg,None,criadores,criadores dietetic adulto dermatosis pienso pa...,111,NaN,18.99€ a 106.38€,4.7,Yahima,None,TD_00010,-20% dto. criadores dietetic adulto dermatosis...,Tiendanimal,None,https://www.tiendanimal.es/perros/pienso-para-...


In [7]:
# ver datos nulos
df_nulos = df_raw.select([
    count(
        when(col(c).isNull(), c)
    ).alias(c)
    for c in df_raw.columns
])

display(df_nulos.toPandas().T.rename(columns={0: "Cantidad de Nulos"}))

,Cantidad de Nulos
_id,0
categoria,0
fecha_captura,0
formato_raw,0
marca,0
nombre_producto,0
opiniones,0
precio_raw,0
rating,0
responsable,0


In [8]:
# Detecta duplicados
duplicados = df_raw.groupBy("sku_id","tienda").count().filter(col("count") > 1)
duplicados.show()
# Eliminar duplicados
df_clean = df_raw.dropDuplicates(["sku_id", "tienda"])

+------+------+-----+
|sku_id|tienda|count|
+------+------+-----+
+------+------+-----+



In [17]:
# Normalización de datos
# Etraccion de precio numérico
from pyspark.sql.functions import (regexp_extract,regexp_replace,when,col,lit)

# Extraer el primer número desde precio_raw
df_clean = df_clean.withColumn("precio_num",regexp_extract(
        col("precio_raw"),r"([0-9]+[.,][0-9]+)",1))

# Reemplazar coma por punto y convertir a número
df_clean = df_clean.withColumn("precio_num",regexp_replace(col("precio_num"),",",".").cast("double"))

# Tiendanimal trabaja en EUR, crear una nueva columna
df_clean = df_clean.withColumn("moneda",lit("EUR"))

# Normalización de moneda hacia CLP
# 1 EUR ≈ 980 CLP

df_clean = df_clean.withColumn("precio_clp",
    when(
        col("moneda") == "EUR",
        col("precio_num") * 980)
        .otherwise(col("precio_num")))

#verificar resultados
df_clean.select(
    "nombre_producto",
    "precio_raw",
    "precio_num",
    "moneda",
    "precio_clp"
).show(10, truncate=False)

+----------------------------------------------------------------------+----------------+----------+------+------------------+
|nombre_producto                                                       |precio_raw      |precio_num|moneda|precio_clp        |
+----------------------------------------------------------------------+----------------+----------+------+------------------+
|acana adult pienso para perros                                        |19.99€ a 195.98€|19.99     |EUR   |19590.199999999997|
|criadores dietetic hipoalergénico adulto pienso para perros           |19.99€ a 123.48€|19.99     |EUR   |19590.199999999997|
|nature's variety no grain adult medium maxi salmón pienso para perros |28.59€ a 123.46€|28.59     |EUR   |28018.2           |
|criadores dietetic adulto obesity pienso para perros                  |18.99€ a 106.38€|18.99     |EUR   |18610.199999999997|
|acana light & fit pienso para perros                                  |21.89€ a 143.84€|21.89     |EUR   |2145

### Paso 5: Creación de variables derivadas

- Variable 1

Precio mínimo:
Desde: 19.99€ a 39.18€
extraer: 19.99

- Variable 2

Precio máximo
extraer:39.18

- Variable 3

Precio promedio: (19.99 + 39.18)/2

- Variable 4

Tiene opiniones: 0 / 1

- Variable 5

Nivel de valoración

Por ejemplo:

rating >= 4.5

→ Excelente

rating >= 4

→ Bueno

etc.

In [9]:
#Crear variables derivadas para análisis

df_clean = df_clean.withColumn(
    "tiene_opiniones",
    when(col("opiniones_num") > 0, 1).otherwise(0)
)

df_clean = df_clean.withColumn(
    "nivel_rating",
    when(col("rating_num") >= 4.5, "Excelente")
    .when(col("rating_num") >= 4.0, "Bueno")
    .when(col("rating_num") >= 3.0, "Regular")
    .otherwise("Sin clasificar")
)

df_clean = df_clean.withColumn(
    "rango_precio",
    when(col("precio_promedio") >= 80, "Alto")
    .when(col("precio_promedio") >= 40, "Medio")
    .otherwise("Bajo")
)

df_clean.select(
    "nombre_producto", "marca", "precio_promedio",
    "precio_kg", "rating_num", "opiniones_num",
    "nivel_rating", "rango_precio"
).show(10, truncate=False)

AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `opiniones_num` cannot be resolved. Did you mean one of the following? [`opiniones`, `pagina`, `precio_raw`, `rating`, `responsable`].;
'Project [_id#0, categoria#1, estado_extraccion#2, fecha_captura#3, formato_raw#4, fuente_tipo#5, marca#6, nombre_producto#7, opiniones#8, pagina#9, precio_raw#10, rating#11, responsable#12, responsable_scraper#13, sku_id#14, texto_crudo#15, tienda#16, tipo_dato#17, url_fuente#18, CASE WHEN ('opiniones_num > 0) THEN 1 ELSE 0 END AS tiene_opiniones#262]
+- Deduplicate [sku_id#14, tienda#16]
   +- RelationV2[_id#0, categoria#1, estado_extraccion#2, fecha_captura#3, formato_raw#4, fuente_tipo#5, marca#6, nombre_producto#7, opiniones#8, pagina#9, precio_raw#10, rating#11, responsable#12, responsable_scraper#13, sku_id#14, texto_crudo#15, tienda#16, tipo_dato#17, url_fuente#18]  MongoTable()


In [20]:
# Extracción de Peso (ingeniería de Variables)

from pyspark.sql.functions import col, regexp_extract, regexp_replace

df_clean = df_clean.withColumn("precio_por_kg_eur", regexp_extract(col("formato_raw"), r"([0-9]+[.,][0-9]+)€\s*el\s*kg", 1))

df_clean = df_clean.withColumn("precio_por_kg_eur", regexp_replace(col("precio_por_kg_eur"), ",", ".").cast("double"))

df_clean = df_clean.withColumn("precio_por_kg_clp", col("precio_por_kg_eur") * 980)

df_clean.select("nombre_producto", "formato_raw", "precio_por_kg_eur", "precio_por_kg_clp").show(20, truncate=False)

+-----------------------------------------------------------------------------+-----------+-----------------+-----------------+
|nombre_producto                                                              |formato_raw|precio_por_kg_eur|precio_por_kg_clp|
+-----------------------------------------------------------------------------+-----------+-----------------+-----------------+
|Acana Adult pienso para perros                                               |5.76€ el kg|5.76             |5644.8           |
|Salvaje Base Adultos Pollo pienso para perros                                |1.27€ el kg|1.27             |1244.6           |
|Nature's Variety No Grain Adult Medium Maxi Salmón pienso para perros        |5.14€ el kg|5.14             |5037.2           |
|Hill's Prescription Diet Food Sensitives z/d pienso para perros              |9.02€ el kg|9.02             |8839.6           |
|Nature's Variety No Grain Adult Medium Maxi Pollo pienso para perros         |4.90€ el kg|4.9          

# Análisis Exploratorio de Datos con Spark (EDA:Exploratory Data Analysis)

In [11]:
# Resumen de variables numericas (Precio y Peso)
df_clean.select("precio_eur", "peso_kg", "rating").describe().show()

AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `precio_eur` cannot be resolved. Did you mean one of the following? [`precio_raw`, `opiniones`, `rating`, `_id`, `pagina`].;
'Project ['precio_eur, 'peso_kg, rating#11]
+- Deduplicate [sku_id#14, tienda#16]
   +- RelationV2[_id#0, categoria#1, estado_extraccion#2, fecha_captura#3, formato_raw#4, fuente_tipo#5, marca#6, nombre_producto#7, opiniones#8, pagina#9, precio_raw#10, rating#11, responsable#12, responsable_scraper#13, sku_id#14, texto_crudo#15, tienda#16, tipo_dato#17, url_fuente#18]  MongoTable()
